recipe/magento2.php

```diff
@@ -219,3 +221,137 @@
+});
+
+desc('Extracts artifact in release path.');
+task('artifact:extract', function () {
+    run('{{bin/tar}} -xzpf {{release_path}}/{{artifact_file}} -C {{release_path}}');
+    run('rm -rf {{release_path}}/{{artifact_file}}');
+});
+
+desc('Provides env.php for build.');
+task('build:prepare-env', function() {
+    $deployEnv = get('deploy_env','app/etc/deploy.php');
+    if (!test('[ -f ./'.$deployEnv.' ]')) {
+        throw new GracefulShutdownException(
+            "No deploy env provided, provide one at app/etc/deploy.php or change location"
+        );
+    }
+    run ('cp '.$deployEnv.' app/etc/env.php');
+});
+
+desc('Clears generated files prior to building.');
+task('build:remove-generated', function() {
+    run('rm -rf generated/*');
+});
+
+desc('Clears the opcache, cache tool required.');
+task('cache:clear:opcache', function() {
+    if ($fpmSocket = get('fpm_socket', '')) {
+        run('{{bin/php}} -f {{cacheToolPath}} opcache:reset --fcgi '.$fpmSocket);
+    }
+});
+
+desc('Builds an artifact.');
+task('artifact:build', function () {
+    if(currentHost()->get('local')) {
+        set('deploy_path', '.');
+        set('release_path', '.');
+        set('current_path', '.');
+        invoke('build:prepare-env');
+        invoke('build:remove-generated');
+        invoke('deploy:vendors');
+        invoke('magento:compile');
+        invoke('magento:deploy:assets');
+        invoke('artifact:package');
+    } else {
+        throw new GracefulShutdownException("Artifact can only be built locally, you provided a non local host");
+    }
+});
+
+desc('Prepares an artifact on the target server');
+task('artifact:prepare', function(){
+    if(currentHost()->get('local')) {
+        throw new GracefulShutdownException("You can only deploy to a non localhost");
+    } else {
+        add('shared_files', get('additional_shared_files') ?? []);
+        add('shared_dirs', get('additional_shared_dirs') ?? []);
+        invoke('deploy:info');
+        invoke('deploy:setup');
+        invoke('deploy:lock');
+        invoke('deploy:release');
+        invoke('artifact:upload');
+        invoke('artifact:extract');
+        invoke('deploy:shared');
+        invoke('deploy:writable');
+    }
+});
+
+desc('Executes the tasks after artifact is released');
+task('artifact:finish', function() {
+    if(currentHost()->get('local')) {
+        throw new GracefulShutdownException("You can only deploy to a non localhost");
+    } else {
+        invoke('magento:cache:flush');
+        invoke('cache:clear:opcache');
+        invoke('deploy:cleanup');
+        invoke('deploy:unlock');
+    }
+});
+
+desc('Actually releases the artifact deployment');
+task('artifact:deploy', function()  {
+
+    if(currentHost()->get('local')) {
+        throw new GracefulShutdownException("You can only deploy to a non localhost");
+    } else {
+        invoke('artifact:prepare');
+
+        invoke('magento:upgrade:db');
+        invoke('magento:config:import');
```

Comment:
We have a problem with all those `invoke()`, apparently they only execute this one task and not anything added via `before()` or `after()`, which makes the recipe hard to extend for custom needs.

Maybe `task()->select()` can be used to only run it on non localhost? If not, is this restriction so important?

Reply:
Me too (just noticed, because I had to add magepack:build :) 
Could the problem possibly be solved by defining a task
```
desc('Forces the tasks to be executed locally and sets the pathes');
task('build:force-local', function () {
    if(currentHost()->get('local')) {
        set('deploy_path', '.');
        set('release_path', '.');
        set('current_path', '.');
    } else {
        throw new GracefulShutdownException("Artifact can only be built locally, you provided a non local host");
    }
});

desc('Builds an artifact.');
task('artifact:build', [
    'build:force-local',
    'build:remove-generated',
    'deploy:vendors',
    'magento:compile',
    'magento:deploy:assets',
    'artifact:package'
]);
```

Reply:
Would also apply to the deployment which must not be local!